### PR TITLE
Replace read/write lock in JarResource to avoid virtual threads pinning

### LINKED ISSUE #42067: App with Infinispan fails after changes in JarFileReference that rarely results in illegal state after changes on ReadWriteLock
### Describe the bug

I have application with Infinispan extension that is deployed to Openshift using the OpenShift extension. Lately I have experienced flakiness when Infinipan library performs service loading on application startup. My colleague @jedla97 discovered that this started happen after https://github.com/quarkusio/quarkus/pull/40942.

### Expected behavior

No race.

### Actual behavior

Application startup fails with exception:

```
03:16:59,364 INFO  [two] 03:16:57,505 Failed to initialize a channel. Closing: [id: 0xd1e053bd]: java.lang.IllegalStateException: The reference counter cannot be negative, found: -1
03:16:59,364 INFO  [two] 	at io.quarkus.bootstrap.runner.JarFileReference.release(JarFileReference.java:58)
03:16:59,364 INFO  [two] 	at io.quarkus.bootstrap.runner.JarFileReference.consumeSharedJarFile(JarFileReference.java:129)
03:16:59,364 INFO  [two] 	at io.quarkus.bootstrap.runner.JarFileReference.withJarFile(JarFileReference.java:107)
03:16:59,364 INFO  [two] 	at io.quarkus.bootstrap.runner.JarResource.getResourceURL(JarResource.java:89)
03:16:59,364 INFO  [two] 	at io.quarkus.bootstrap.runner.RunnerClassLoader.findResources(RunnerClassLoader.java:265)
03:16:59,364 INFO  [two] 	at java.base/java.lang.ClassLoader.getResources(ClassLoader.java:1473)
03:16:59,364 INFO  [two] 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1203)
03:16:59,364 INFO  [two] 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1228)
03:16:59,364 INFO  [two] 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
03:16:59,364 INFO  [two] 	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
03:16:59,364 INFO  [two] 	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
03:16:59,364 INFO  [two] 	at org.infinispan.commons.util.SaslUtils.getFactories(SaslUtils.java:58)
03:16:59,365 INFO  [two] 	at org.infinispan.commons.util.SaslUtils.getSaslClientFactories(SaslUtils.java:49)
03:16:59,365 INFO  [two] 	at org.infinispan.client.hotrod.impl.transport.netty.ChannelInitializer.getSaslClientFactory(ChannelInitializer.java:227)
03:16:59,365 INFO  [two] 	at org.infinispan.client.hotrod.impl.transport.netty.ChannelInitializer.initAuthentication(ChannelInitializer.java:204)
03:16:59,365 INFO  [two] 	at org.infinispan.client.hotrod.impl.transport.netty.ChannelInitializer.initChannel(ChannelInitializer.java:110)
03:16:59,365 INFO  [two] 	at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:129)
03:16:59,365 INFO  [two] 	at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:112)
03:16:59,365 INFO  [two] 	at io.netty.channel.AbstractChannelHandlerContext.callHandlerAdded(AbstractChannelHandlerContext.java:1130)
03:16:59,365 INFO  [two] 	at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:608)
03:16:59,365 INFO  [two] 	at io.netty.channel.DefaultChannelPipeline.access$100(DefaultChannelPipeline.java:45)
03:16:59,365 INFO  [two] 	at io.netty.channel.DefaultChannelPipeline$PendingHandlerAddedTask.execute(DefaultChannelPipeline.java:1460)
03:16:59,365 INFO  [two] 	at io.netty.channel.DefaultChannelPipeline.callHandlerAddedForAllHandlers(DefaultChannelPipeline.java:1114)
03:16:59,365 INFO  [two] 	at io.netty.channel.DefaultChannelPipeline.invokeHandlerAddedIfNeeded(DefaultChannelPipeline.java:649)
03:16:59,365 INFO  [two] 	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:513)
03:16:59,365 INFO  [two] 	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:428)
03:16:59,365 INFO  [two] 	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:485)
03:16:59,365 INFO  [two] 	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
03:16:59,365 INFO  [two] 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
03:16:59,365 INFO  [two] 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
03:16:59,365 INFO  [two] 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
03:16:59,366 INFO  [two] 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
03:16:59,366 INFO  [two] 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
03:16:59,366 INFO  [two] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
03:16:59,366 INFO  [two] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
03:16:59,366 INFO  [two] 	at java.base/java.lang.Thread.run(Thread.java:840)
03:16:59,366 INFO  [two] 03:16:57,517 ISPN004007: Exception encountered. Retry 0 out of 0: io.netty.channel.StacklessClosedChannelException
03:16:59,366 INFO  [two] 	at io.netty.channel.AbstractChannel$AbstractUnsafe.ensureOpen(ChannelPromise)(Unknown Source)
03:16:59,366 INFO  [two] 03:16:57,517 ISPN004007: Exception encountered. Retry 0 out of 0: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: ts-qhdkqxqyjo-infinispan-cluster.datagrid-cluster.svc.cluster.local/172.30.48.250:11222
03:16:59,366 INFO  [two] Caused by: java.net.ConnectException: Connection refused
03:16:59,366 INFO  [two] 	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
03:16:59,366 INFO  [two] 	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
03:16:59,366 INFO  [two] 	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:946)
03:16:59,366 INFO  [two] 	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:336)
03:16:59,366 INFO  [two] 	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:339)
03:16:59,366 INFO  [two] 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:776)
03:16:59,366 INFO  [two] 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
03:16:59,366 INFO  [two] 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
03:16:59,366 INFO  [two] 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
03:16:59,367 INFO  [two] 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
03:16:59,367 INFO  [two] 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
03:16:59,367 INFO  [two] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
03:16:59,367 INFO  [two] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
03:16:59,367 INFO  [two] 	at java.base/java.lang.Thread.run(Thread.java:840)
```

### How to Reproduce?

Reproducer:

0. have OpenShift cluster up and running, be logged in there
1. `git clone git@github.com:quarkus-qe/quarkus-test-suite.git`
2. `git checkout e4056763aecf9a048b89a0e4f2b5ae0e9a98da25` (that just because I don't know if I won't disable the test, you likely can skip this step)
3. `quarkus-test-suite/infinispan-client`
4. `mvn clean verify -Dopenshift -Dit.test=OperatorOpenShiftInfinispanCountersIT -Doc.reruns=0`

But I'll be honest with you, our Jenkins has **by far** more success in reproducing it than I do on my workstation. I have never actually seen this exception outside of Jenkins logs.

### Output of `uname -a` or `ver`

Fedora 40

### Output of `java -version`

17.0.10, vendor: Red Hat, Inc., runtime: /qa/tools/opt/x86_64/java-17-openjdk-17.0.10.0.7-1.portable.jdk.el.x86_64

### Quarkus version or git rev

999-SNAPSHOT

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by mariofusco**
This pull request fixes the [race condition](https://github.com/quarkusio/quarkus/issues/42067) present in [this commit](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f) that, because of it, had to be reverted.

/cc @geoand @gsmet @franz1981 @jedla97 @michalvavrik 

Fixes: #42067

The nature of the fix is described at https://github.com/quarkusio/quarkus/pull/42139#pullrequestreview-2199296948

### CODE REVIEW

File: independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        try {
+            return CompletableFuture.completedFuture(new JarFileReference(JarFiles.create(jarResource.jarPath.toFile())));
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to open " + jarResource.jarPath, e);
+        }
+    }
+
+    private static JarFileReference asyncLoadAcquiredJarFile(JarResource jarResource) {
+        CompletableFuture<JarFileReference> newJarRefFuture = new CompletableFuture<>();
+        CompletableFuture<JarFileReference> existingJarRefFuture = null;
```
Comment (mariofusco) [inline independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java:158]: The introduction of this variable is the actual fix. The old algorithm [wrongly reused](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f#diff-f20997e90aae1d2fe046b512defb1a3e14215b7a834177d43369542bae99a7f6R168) the `newJarFileRef` one. This means that in the unfortunate case when a thread was [about to close the old jar](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f#diff-f20997e90aae1d2fe046b512defb1a3e14215b7a834177d43369542bae99a7f6R60), but [before setting the AtomicReference to null](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f#diff-f20997e90aae1d2fe046b512defb1a3e14215b7a834177d43369542bae99a7f6R67), another thread could fail the `compareAndSet(null)` check here, take at the end of the loop from the `AtomicReference` the old and no longer valid `CompletableFuture` and in subsequent loop (after that the close finally managed to set the `AtomicReference` to null) mistakenly set it back into the same `AtomicReference`.

**Comment by geoand** [comment]
Thanks for the fix @mariofusco !

@jedla97 thanks for signing up to test the fix! Hopefully @famod can do so as well

**Comment by jedla97** [comment]
@mariofusco Big thanks for the fix as I believe it was hard to find it. I tested it (run two reproducer multiple times) and the race condition didn't appear. So from my POV the original issue is fixed (hopefully we won't see any different :pray: ).

**Comment by franz1981** [comment]
It would be great if there is a test which reproduce it too, with the changes to fix

**Comment by famod** [comment]
Sorry, I am unable to test this ATM. On my work machine, I'm getting a puzzling dependency resolution issue while building, it's related to gradle-tooling.

**Comment by mariofusco** [comment]
> It would be great if there is a test which reproduce it too, with the changes to fix

I did a few more checks on this fix and also added a proper unit test. The unit test reproduces exactly the same situation that caused the `IllegalStateException` in the former version of this class loader as I described it [here](https://github.com/quarkusio/quarkus/pull/42139#discussion_r1691438739). Of course, being a race condition, the test is not deterministic and it fails with the older version extremely seldomly. However it's descriptive of the problem and I'm afraid it will be impossible to make it reproducible in a more reliable way. 

@geoand @franz1981 the pull request has been already rebased and squashed as usual, so it should be ready to be merged. Please let me know if this is enough or you see any room for further improvements.

**Review by franz1981 — APPROVED** [review]
LGTM well done as usual Mario!

**Review by geoand — APPROVED** [review]
Thanks a lot!

**Comment by mariofusco** [comment]
> ## Failing Jobs
> Status 	Name 	Step 	Failures 	Logs 	Raw logs 	Build scan
> ✖ 	JVM Tests - JDK 17 	`Build` 	[Failures](#user-content-test-failures-job-28064977823) 	[Logs](https://github.com/quarkusio/quarkus/actions/runs/10149362227/job/28064977823) 	[Raw logs](https://github.com/quarkusio/quarkus/commit/b2b793c9e7bf6d5cebe860fe556af2e2aa5a094c/checks/28064977823/logs) 	[🔍](https://ge.quarkus.io/s/nvcms2tcq3nba)
> ✖ 	JVM Tests - JDK 17 Windows 	`Build` 	[Failures](#user-content-test-failures-job-28064978975) 	[Logs](https://github.com/quarkusio/quarkus/actions/runs/10149362227/job/28064978975) 	[Raw logs](https://github.com/quarkusio/quarkus/commit/b2b793c9e7bf6d5cebe860fe556af2e2aa5a094c/checks/28064978975/logs) 	[🔍](https://ge.quarkus.io/s/2e7oemre4jm26)

These failures are caused by the fact that I compiled the minimal jars that I used for this test with Java 21, so they cannot be loaded with Java 17.

`java.lang.UnsupportedClassVersionError: org/simple/SimplePojo1 has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 61.0`

I will recompile those jars with Java 17 and replace them in this pull request immediately.

**Comment by mariofusco** [comment]
> I will recompile those jars with Java 17 and replace them in this pull request immediately.

Done.

**Comment by mariofusco** [comment]
Is this ready to be merged? Since the 3.13.0 is already out there shouldn't be any reason to hold it, right?

**Comment by geoand** [comment]
Correct

**Comment by geoand** [comment]
Let's keep an eye out on weird failures as we did with the previous iteration.

**Comment by mariofusco** [comment]
> Let's keep an eye out on weird failures as we did with the previous iteration.

Thank you for merging. Sure, please ping me as soon as you will notice something wrong.

**Comment by geoand** [comment]
Will do! Thanks again for tackling this!

If anything comes up, it will be by QE running their testsuite over the next few days. Otherwise if there is an issue, reports from users will start coming in when `3.14.0.CR1` is released.

**Pull request merged**